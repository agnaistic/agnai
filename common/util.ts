import { AppSchema } from './types/schema'

export function replace<T extends { _id: string }>(id: string, list: T[], item: Partial<T>) {
  return list.map((li) => (li._id === id ? { ...li, ...item } : li))
}

export function findOne<T extends { _id: string }>(id: string, list: T[]): T | void {
  for (const item of list) {
    if (item._id === id) return item
  }
}

export function toArray<T>(values?: T | T[]): T[] {
  if (values === undefined) return []
  if (Array.isArray(values)) return values
  return [values]
}

export function distinct<T extends { _id: string }>(values: T[]) {
  const set = new Set<string>()
  const next: T[] = []

  for (const item of values) {
    if (set.has(item._id)) continue
    next.push(item)
    set.add(item._id)
  }

  return next
}
export function wait(secs: number) {
  return new Promise((res) => setTimeout(res, secs * 1000))
}

export function waitMs(ms: number) {
  return new Promise((res) => setTimeout(res, ms))
}

export function isObject(val: unknown) {
  return Object.prototype.toString.call(val) === '[object Object]'
}

export function cycleArray<T>(arr: T[], shiftAmount: number) {
  const effectiveShift = shiftAmount % arr.length
  return [...arr.slice(effectiveShift), ...arr.slice(0, effectiveShift)]
}

export const dateStringFromUnix = (unixTimestamp: number): string =>
  new Date(unixTimestamp).toISOString()

export function uniqBy<T, U>(arr: T[], accessor: (el: T) => U) {
  return arr.flatMap((el, i) => (arr.slice(0, i).map(accessor).includes(accessor(el)) ? [] : [el]))
}

export function toDuration(valueSecs: number, full?: boolean) {
  if (valueSecs === 0) {
    if (full) return '0s'
    return '0 seconds'
  }
  const {
    duration: [days, hours, minutes, seconds],
  } = toRawDuration(valueSecs)

  if (full) {
    return [`${days}d`, `${hours}h`, `${minutes}m`, `${seconds}s`]
      .filter((time) => !time.startsWith('0'))
      .join(':')
  }

  if (days) {
    return `${days} ${days > 1 ? 'days' : 'day'}`
  }

  if (hours) {
    return `${hours} ${hours > 1 ? 'hours' : 'hour'}`
  }

  if (minutes) {
    return `${minutes} ${minutes > 1 ? 'mins' : 'min'}`
  }

  return `${seconds} ${seconds > 1 ? 'seconds' : 'second'}`
}

export function elapsedSince(date: string | Date, offsetMs: number = 0) {
  if (!date) return 'unknown'
  const diff = Date.now() - new Date(date).valueOf() + offsetMs
  const elapsed = Math.floor(diff / 1000)
  if (elapsed < 60) return 'a moment'
  return toDuration(Math.floor(elapsed))
}

const ONE_HOUR = 3600
const ONE_DAY = 86400

type Days = number
type Hours = number
type Minutes = number
type Seconds = number

type Duration = [Days, Hours, Minutes, Seconds]

function toRawDuration(valueSecs: number) {
  const days = Math.floor(valueSecs / ONE_DAY)
  const hours = Math.floor(valueSecs / ONE_HOUR) % 24
  const mins = Math.floor(valueSecs / 60) % 60
  const secs = Math.ceil(valueSecs % 60)

  return {
    duration: [days, hours, mins, secs] as Duration,
    seconds: valueSecs,
    text: valueSecs <= 0 ? 'now' : `${days}d:${hours}h:${mins}m:${secs}s`.replace('0d:', ''),
  }
}

/**
 * Remove leading tabs and whitespace on each line from template strings
 * Also 'trim' the text before returning
 */
export function neat(params: TemplateStringsArray, ...rest: string[]) {
  let str = ''
  for (let i = 0; i < params.length; i++) {
    str += params[i] + (rest[i] || '')
  }

  return str
    .split('\n')
    .map((line) => line.replace(/^[\t ]+/g, ''))
    .join('\n')
    .trim()
}

const end = `."'*!?)}]\``.split('')
export function trimSentence(text: string) {
  const last = end.reduce((last, char) => {
    const index = text.lastIndexOf(char)
    return index > last ? index : last
  }, -1)

  if (last === -1) return text
  return text.slice(0, last + 1)
}

export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, '')
    .replace(/[\s_-]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

// https://stackoverflow.com/a/3561711
export function escapeRegex(string: string) {
  return string.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
}

export function getBotName(
  chat: AppSchema.Chat,
  msg: AppSchema.ChatMessage,
  chars: Record<string, AppSchema.Character>,
  replyAs: AppSchema.Character,
  main: AppSchema.Character,
  impersonate?: AppSchema.Character
) {
  if (!msg.characterId) return replyAs?.name || main.name
  if (msg.characterId.startsWith('temp-')) {
    const temp = chat.tempCharacters?.[msg.characterId]
    if (!temp) return main.name
    return temp.name
  }

  const char =
    msg.characterId && impersonate?._id === msg.characterId ? impersonate : chars[msg.characterId]

  if (!char) {
    return main.name
  }

  return char.name
}

export function eventGenerator<T = any>() {
  const queue: Array<Promise<any>> = []
  let done = false
  let signal = false

  const stream = (async function* () {
    while (!done) {
      await waitMs(1)
      const promise = queue.shift()
      if (!promise) {
        if (signal) {
          done = true
        }
        continue
      }

      const result = await promise
      yield result
    }
  })() as AsyncGenerator<T, T>

  return {
    stream,
    push: (value: T) => {
      if (done) return
      queue.push(Promise.resolve(value))
    },
    done: () => {
      if (done) return
      signal = true
    },
  }
}

/**
 * Will automatically clamp to `-max -> max` if `min` is not provided
 */
export function clamp(toClamp: number, max: number, min?: number) {
  return Math.max(Math.min(toClamp, max), min ?? -max)
}

export function now() {
  return new Date().toISOString()
}
