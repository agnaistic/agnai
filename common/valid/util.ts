import { OptionalPrimitive, Primitive, Validator } from './types'

export function isPrimitive(value: any): value is Primitive {
  return (
    typeof value === 'string' &&
    (value === 'string' ||
      value === 'boolean' ||
      value === 'number' ||
      value === 'any' ||
      value === 'unknown')
  )
}

export function isOptionalPrimitive(value: any): value is OptionalPrimitive {
  return (
    value === 'string?' ||
    value === 'boolean?' ||
    value === 'number?' ||
    value === 'any?' ||
    value === 'unknown?'
  )
}

export function isTuplePrimitive(value: any): value is [Primitive] | readonly [Primitive] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 1) return false
  if (!isPrimitive(value[0])) return false
  return true
}

export function isTupleOptional(
  value: any
): value is [OptionalPrimitive] | readonly [OptionalPrimitive] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 1) return false
  if (!isOptionalPrimitive(value[0])) return false
  return true
}

export function isOptionalArray(value: any): value is [Validator, '?'] | readonly [Validator, '?'] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 2) return false
  return value[1] === '?'
}

export function isObjectOptional(value: any): value is Validator & { '?': any } {
  return typeof value === 'object' && '?' in value && !!value['?']
}

export function isTupleBody(
  value: any
): value is [Validator] | readonly [Validator] | [Validator, '?'] | readonly [Validator, '?'] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 1 && value.length !== 2) return false
  if (value.length === 2) return typeof value[0] === 'object' && value[1] === '?'
  return typeof value[0] === 'object'
}

export function isOptionalUnion<T extends string>(
  value: any
): value is T[] | readonly string[] | undefined {
  if (!Array.isArray(value)) return false
  if (value.length < 1) return false
  if (isPrimitive(value[0])) return false
  if (value.includes(null)) return true
  return false
}

export function isUnion<T extends string>(value: any): value is [T] | string[] | readonly string[] {
  if (Array.isArray(value) === false) return false
  if (value.length < 1) return false
  if (isPrimitive(value[0])) return false

  return true
}
