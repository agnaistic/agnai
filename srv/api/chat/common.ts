import { customSettings } from '../../config'
import { AppSchema } from '../../../common/types/schema'

export const personaValidator = {
  kind: ['wpp', 'sbf', 'boostyle', 'text'],
  attributes: 'any',
} as const

export function extractActions(text: string) {
  const [split, rawActions] = text.split('ACTIONS:')
  if (!rawActions || !rawActions.trim()) {
    let responses: string[] = []
    const actions: AppSchema.ChatMessage['actions'] = []
    const splits = text.split('\n')

    for (const split of splits) {
      const [emote, action] = split.split('->')
      if (!action) {
        responses.push(emote)
        continue
      }

      actions.push({ emote: emote.trim().replace('{', '').replace('}', ''), action: action.trim() })
    }

    return { text: responses.join('\n'), actions }
  }

  const actions = rawActions
    .split('\n')
    .filter((line) => !!line.trim())
    .map((line) => line.trim().split('-'))
    .map(([emote, action]) => ({ emote, action }))

  return { text: split.trim(), actions }
}

export function trimResponseV2(
  generated: string,
  char: AppSchema.Character,
  members: AppSchema.Profile[],
  bots: Record<string, AppSchema.Character> | undefined,
  endTokens: string[] = []
) {
  const allEndTokens = getEndTokens(null, members)

  generated = generated.split(`${char.name} :`).join(`${char.name}:`)

  for (const member of members) {
    if (!member.handle) continue
    generated = generated.split(`${member.handle} :`).join(`${member.handle}:`)
  }

  if (bots) {
    for (const bot of Object.values(bots)) {
      if (!bot) continue
      if (bot?._id === char._id) continue
      endTokens.push(`${bot.name}:`)
    }
  }

  let index = -1
  let trimmed = allEndTokens.concat(...endTokens).reduce((prev, endToken) => {
    const idx = generated.indexOf(endToken)

    if (idx === -1) return prev

    const text = generated.slice(0, idx)
    if (index === -1 || idx < index) {
      index = idx
      return text
    }

    return prev
  }, '')

  if (index === -1) {
    return sanitise(generated.split(`${char.name}:`).join(''))
  }

  return sanitise(trimmed.split(`${char.name}:`).join(''))
}

export function sanitiseAndTrim(
  text: string,
  prompt: string,
  char: AppSchema.Character,
  characters: Record<string, AppSchema.Character> | undefined,
  members: AppSchema.Profile[]
) {
  const parsed = sanitise(text.replace(prompt, ''))
  const trimmed = trimResponseV2(parsed, char, members, characters, ['END_OF_DIALOG'])
    .split(`${char.name}:`)
    .join('')
  return trimmed || parsed
}

export function getEndTokens(
  char: AppSchema.Character | null,
  members: AppSchema.Profile[],
  endTokens: string[] = []
) {
  const baseEndTokens = ['END_OF_DIALOG', '<END>'].concat(endTokens)

  if (customSettings.baseEndTokens) {
    baseEndTokens.push(...customSettings.baseEndTokens)
  }

  if (char) {
    baseEndTokens.push(`${char.name}:`, `${char.name} :`)
  }

  for (const member of members) {
    baseEndTokens.push(`${member.handle}:`, `${member.handle} :`)
  }

  const uniqueTokens = Array.from(new Set(baseEndTokens))
  return uniqueTokens
}

export function joinParts(parts: string[]) {
  return parts.map(sanitise).join(' ').trim()
}

export function sanitise(generated: string) {
  // If want to support code blocks we need to remove the excess whitespace removal as it breaks indents
  return generated.trim()
}

export function normalizeUrl(url: string, defaultUrl?: string) {
  if (!url) return defaultUrl
  return url.replace(/\/$/, '')
}
