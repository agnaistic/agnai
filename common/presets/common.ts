import { neat } from '../util'

export const gaslights = {
  alpaca: neat`
  Below is an instruction that describes a task. Write a response that appropriately completes the request.
  
  Write {{char}}'s next reply in a fictional roleplay chat between {{#each bot}}{{.name}}, {{/each}}{{char}}.
  
  {{char}}'s Persona: {{personality}}
  
  This scenario of the conversation: {{scenario}}
  
  This is how {{char}} should talk: {{example_dialogue}}
  
  Then the roleplay chat between {{#each bot}}{{.name}}, {{/each}}{{char}} begins.
  
  {{#each msg}}{{#if .isbot}}### Response:\n{{.name}}: {{.msg}}{{/if}}{{#if .isuser}}### Instruction:\n{{.name}}: {{.msg}}{{/if}}
  {{/each}}
  {{ujb}}
  ### Response:
  {{post}}
  `,
}
