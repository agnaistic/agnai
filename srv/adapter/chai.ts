import needle from 'needle'
import { config } from '../config'
import { trimResponse, trimResponseV2 } from '../api/chat/common'
import { ModelAdapter } from './type'

export const handleChai: ModelAdapter = async function* ({ char, members, prompt, settings, log }) {
  if (!config.chai.url) {
    yield { error: 'Chai URL not set' }
    return
  }

  if (config.chai.uid === 'empty' || config.chai.key === 'empty') {
    yield { error: 'Chai is not configured' }
    return
  }

  const body = { ...settings, text: prompt }

  log.debug(body, 'Chai payload')

  const response = await needle('post', `${config.chai.url}/generate/gptj`, body, {
    json: true,
    timeout: 3000,
    response_timeout: 10000,
    headers: {
      'user-agent':
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36',
      origin: 'https://chai.ml',
      developer_uid: config.chai.uid,
      developer_key: config.chai.key,
    },
  }).catch((err) => ({ err }))

  if ('err' in response) {
    yield { error: response.err.message }
    return
  }

  const status = response.statusCode || 0
  if (status >= 500) {
    log.error(response.body, 'Chai request failed')
    yield { error: `Chai internal server error: ` }
    return
  }

  if (status >= 400) {
    log.error(response.body, 'Chai request failed')
    yield { error: 'Chai request error' }
    return
  }

  const username = 'You'
  const endTokens = [
    `Me:`,
    `Me :`,
    `${username}:`,
    `${char.name}:`,
    `${username} :`,
    `${char.name} :`,
    'END_OF_DIALOG',
  ]

  const trimmed = trimResponseV2(response.body.data, char, members, endTokens)
  yield trimmed || response.body.data
}
