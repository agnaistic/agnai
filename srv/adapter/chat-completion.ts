import needle from 'needle'
import { publishOne } from '../api/ws/handle'
import { getTokenCounter } from '../tokenize'
import { needleToSSE } from './stream'
import { AdapterProps } from './type'
import { OPENAI_MODELS } from '/common/adapters'
import { defaultPresets } from '/common/default-preset'
import { IMAGE_SUMMARY_PROMPT } from '/common/image'
import {
  BOT_REPLACE,
  SAMPLE_CHAT_MARKER,
  SELF_REPLACE,
  ensureValidTemplate,
  injectPlaceholders,
} from '/common/prompt'
import { AppSchema } from '/common/types'
import { escapeRegex } from '/common/util'
import { AppLog } from '../logger'

type Role = 'user' | 'assistant' | 'system'
type CompletionItem = { role: Role; content: string; name?: string }

type SplitSampleChatProps = {
  sampleChat: string
  char: string
  sender: string
  budget?: number
}

type CompletionContent<T> = Array<{ finish_reason: string; index: number } & ({ text: string } | T)>
type Inference = { message: { content: string; role: Role } }
type AsyncDelta = { delta: Partial<Inference['message']> }

type Completion<T = Inference> = {
  id: string
  created: number
  model: string
  object: string
  choices: CompletionContent<T>
  error?: { message: string }
}

type CompletionGenerator = (
  userId: string,
  url: string,
  headers: Record<string, string | string[] | number>,
  body: any,
  service: string,
  log: AppLog
) => AsyncGenerator<
  { error: string } | { error?: undefined; token: string },
  Completion | undefined
>

// We only ever use the OpenAI gpt-3 encoder
// Don't bother passing it around since we know this already
const encoder = () => getTokenCounter('openai', OPENAI_MODELS.Turbo)

const sampleChatMarkerCompletionItem: CompletionItem = {
  role: 'system',
  content: SAMPLE_CHAT_MARKER.replace('System: ', ''),
}

/**
 * Yields individual tokens as OpenAI sends them, and ultimately returns a full completion object
 * once the stream is finished.
 */
export const streamCompletion: CompletionGenerator = async function* (
  userId,
  url,
  headers,
  body,
  service,
  log
) {
  const resp = needle.post(url, JSON.stringify(body), {
    parse: false,
    headers: {
      ...headers,
      Accept: 'text/event-stream',
    },
  })

  const tokens = []
  let meta = { id: '', created: 0, model: '', object: '', finish_reason: '', index: 0 }

  try {
    const events = needleToSSE(resp)
    for await (const event of events) {
      if (!event.data) {
        continue
      }

      if (event.type === 'ping') {
        continue
      }

      if (event.data === '[DONE]') {
        break
      }

      const parsed: Completion<AsyncDelta> = JSON.parse(event.data)
      const { choices, ...evt } = parsed
      if (!choices || !choices[0]) {
        log.warn({ sse: event }, `[${service}] Received invalid SSE during stream`)

        const message = evt.error?.message
          ? `${service} interrupted the response: ${evt.error.message}`
          : `${service} interrupted the response`

        if (!tokens.length) {
          yield { error: message }
          return
        }

        publishOne(userId, { type: 'notification', level: 'warn', message })
        break
      }

      const { finish_reason, index, ...choice } = choices[0]

      meta = { ...evt, finish_reason, index }

      if ('text' in choice) {
        const token = choice.text
        tokens.push(token)
        yield { token }
      } else if ('delta' in choice && choice.delta.content) {
        const token = choice.delta.content
        tokens.push(token)
        yield { token }
      }
    }
  } catch (err: any) {
    log.error({ err }, `${service} streaming request failed`)
    yield { error: `${service} streaming request failed: ${err.message}` }
    return
  }

  return {
    id: meta.id,
    created: meta.created,
    model: meta.model,
    object: meta.object,
    choices: [
      {
        finish_reason: meta.finish_reason,
        index: meta.index,
        text: tokens.join(''),
      },
    ],
  }
}

export function toChatCompletionPayload(opts: AdapterProps, maxTokens: number): CompletionItem[] {
  if (opts.kind === 'plain') {
    return [{ role: 'system', content: opts.prompt }]
  }

  const { lines, parts, gen, replyAs } = opts

  const messages: CompletionItem[] = []
  const history: CompletionItem[] = []

  const handle = opts.impersonate?.name || opts.sender?.handle || 'You'
  const gaslight = injectPlaceholders(
    ensureValidTemplate(gen.gaslight || defaultPresets.openai.gaslight, opts.parts, [
      'history',
      'post',
    ]),
    {
      opts,
      parts,
      lastMessage: opts.lastMessage,
      characters: opts.characters || {},
      encoder: encoder(),
    }
  )

  messages.push({ role: 'system', content: gaslight })

  const all = []

  let maxBudget = (gen.maxContextLength || defaultPresets.openai.maxContextLength) - maxTokens
  let tokens = encoder()(gaslight)

  if (lines) {
    all.push(...lines)
  }

  // Append 'postamble' and system prompt (ujb)
  const post = getPostInstruction(opts, messages)
  if (post) {
    post.content = injectPlaceholders(post.content, {
      opts,
      parts: opts.parts,
      lastMessage: opts.lastMessage,
      characters: opts.characters || {},
      encoder: encoder(),
    })
    tokens += encoder()(post.content)
    history.push(post)
  }

  const examplePos = all.findIndex((l) => l.includes(SAMPLE_CHAT_MARKER))

  for (let i = all.length - 1; i >= 0; i--) {
    const line = all[i]

    const obj: CompletionItem = {
      role: 'assistant',
      content: line.trim().replace(BOT_REPLACE, replyAs.name).replace(SELF_REPLACE, handle),
    }

    const isSystem = line.startsWith('System:')
    const isUser = line.startsWith(handle)
    const isBot = !isUser && !isSystem

    if (i === examplePos) {
      const { additions, consumed } = splitSampleChat({
        budget: maxBudget - tokens,
        sampleChat: obj.content,
        char: replyAs.name,
        sender: handle,
      })

      if (tokens + consumed > maxBudget) continue
      history.push(...additions.reverse())
      tokens += consumed
      continue
    } else if (isBot) {
    } else if (line === '<START>') {
      obj.role = sampleChatMarkerCompletionItem.role
      obj.content = sampleChatMarkerCompletionItem.content
    } else if (isSystem) {
      obj.role = 'system'
      obj.content = obj.content.replace('System:', '').trim()
    } else {
      obj.role = 'user'
    }

    const length = encoder()(obj.content)
    if (tokens + length > maxBudget) break
    tokens += length
    history.push(obj)
  }

  return messages.concat(history.reverse())
}

export function splitSampleChat(opts: SplitSampleChatProps) {
  const { sampleChat, char, sender, budget } = opts
  const regex = new RegExp(
    `(?<=\\n)(?=${escapeRegex(char)}:|${escapeRegex(sender)}:|System:|<start>)`,
    'gi'
  )
  const additions: CompletionItem[] = []
  let tokens = 0

  for (const chat of sampleChat.replace(/\r\n/g, '\n').split(regex)) {
    const trimmed = chat.trim()
    if (!trimmed) continue

    // if the msg starts with <start> we consider everything between
    // <start> and the next placeholder a system message
    if (trimmed.toLowerCase().startsWith('<start>')) {
      const afterStart = trimmed.slice(7).trim()
      additions.push(sampleChatMarkerCompletionItem)
      tokens += encoder()(sampleChatMarkerCompletionItem.content)
      if (afterStart) {
        additions.push({ role: 'system' as const, content: afterStart })
        tokens += encoder()(afterStart)
      }
      continue
    }

    const sample = trimmed.toLowerCase().startsWith('system:') ? trimmed.slice(7).trim() : trimmed
    const role = sample.startsWith(char + ':')
      ? 'assistant'
      : sample.startsWith(sender + ':')
      ? 'user'
      : 'system'

    const msg: CompletionItem = {
      role: role,
      content: sample.replace(BOT_REPLACE, char).replace(SELF_REPLACE, sender),
    }

    const length = encoder()(msg.content)
    if (budget && tokens + length > budget) break

    additions.push(msg)
    tokens += length
  }

  return { additions, consumed: tokens }
}

function getPostInstruction(
  opts: AdapterProps,
  messages: CompletionItem[]
): CompletionItem | undefined {
  let prefix = opts.parts.ujb ? `${opts.parts.ujb}\n\n` : ''

  prefix = injectPlaceholders(prefix, {
    opts,
    parts: opts.parts,
    lastMessage: opts.lastMessage,
    characters: opts.characters || {},
    encoder: encoder(),
  })

  switch (opts.kind) {
    // These cases should never reach here
    case 'plain':
    case 'ooc': {
      return
    }

    case 'continue':
      return { role: 'system', content: `${prefix}Continue ${opts.replyAs.name}'s response` }

    case 'summary': {
      let content = opts.user.images?.summaryPrompt || IMAGE_SUMMARY_PROMPT.openai

      if (!content.startsWith('(')) content = '(' + content
      if (!content.endsWith(')')) content = content + ')'

      const looks = Object.values(opts.characters || {})
        .map(getCharLooks)
        .filter((v) => !!v)
        .join('\n')

      if (looks) {
        messages[0].content += '\n' + looks
      }
      return { role: 'user', content }
    }

    case 'self':
      return {
        role: 'system',
        content: `${prefix}Respond as ${opts.impersonate?.name || opts.sender?.handle || 'You'}`,
      }

    case 'retry':
    case 'send':
    case 'request': {
      return { role: 'system', content: `${prefix}Respond as ${opts.replyAs.name}` }
    }
  }
}

function getCharLooks(char: AppSchema.Character) {
  if (char.persona?.kind === 'text') return

  const visuals = [
    char.persona?.attributes?.looks || '',
    char.persona?.attributes?.appearance || '',
  ].filter((v) => !!v)

  if (!visuals.length) return
  return `${char.name}'s appearance: ${visuals.join(', ')}`
}
