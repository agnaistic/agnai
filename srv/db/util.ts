import crypto from 'crypto'
import bcrypt from 'bcryptjs'
import { config } from '../config'
import { AppSchema } from '../../common/types/schema'
import { defaultChars } from '../../common/characters'

const ALGO = 'aes-192-cbc'
const KEY = crypto.scryptSync(config.jwtSecret, 'salt', 24)

export function now() {
  return new Date().toISOString()
}

export async function encryptPassword(value: string) {
  const salt = await bcrypt.genSalt()
  const hash = await bcrypt.hash(value, salt)

  return hash
}

export function encryptText(text: string) {
  const iv = crypto.randomBytes(16)
  const cipher = crypto.createCipheriv(ALGO, KEY, iv)
  const encrypted = cipher.update(text, 'utf8', 'hex')

  return [encrypted + cipher.final('hex'), Buffer.from(iv).toString('hex')].join('|')
}

export function decryptText(text: string) {
  const [encrypted, iv] = text.split('|')
  if (!iv) throw new Error('IV not found')
  const decipher = crypto.createDecipheriv(ALGO, KEY, Buffer.from(iv, 'hex'))
  return decipher.update(encrypted, 'hex', 'utf8') + decipher.final('utf8')
}

export const STARTER_CHARACTER: AppSchema.Character = {
  _id: '',
  userId: '',
  kind: 'character',
  createdAt: '',
  updatedAt: '',
  favorite: false,
  visualType: 'avatar',
  ...defaultChars.Robot,
}

export async function wait(ms: number) {
  return new Promise((res) => setTimeout(res, ms))
}
