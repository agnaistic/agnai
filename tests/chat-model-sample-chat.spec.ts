import { expect } from 'chai'
import './init'
import { splitSampleChat } from '/srv/adapter/chat-completion'
import { reset } from './util'
import { neat } from '/common/util'

describe('Chat Completion Example Dialogue::', () => {
  before(reset)

  it('should properly convert <START> (case insensitive) and split basic messages', () => {
    const input = neat`<STaRT>
      Sam: hey there Vader!
      Vader: hi Sam!`
    const output = testInput(input)
    expect(output).toMatchSnapshot()
  })

  it('should properly understand the first conversation even if it doesnt begin with <START>', () => {
    const input = neat`Sam: hey there
      Vader: hi!`
    const output = testInput(input)
    expect(output).toMatchSnapshot()
  })

  it('will interpret everything up to the first name+colon (or <start>) as a system message if the sampleChat string doesnt begin with a name+colon', () => {
    const input = neat`Vader is nice.
      Sam: hey
      Vader: hi!`
    const output = testInput(input)
    expect(output).toMatchSnapshot()

    const inputWithStart = neat`Vader is nice.
      <START>
      Sam: hey
      Vader: hi!`
    const outputWithStart = testInput(inputWithStart)
    expect(outputWithStart).toMatchSnapshot()
  })

  it('should understand when multiple <Start> are used in the sample chat and turn them all into the defined System message saying a new conversation has started', () => {
    const input = neat`{{char}} is nice.
      <START>
      Sam: hey
      Vader: hi!
      <START>
      Vader: bye Sam
      Sam: byebye Vader`
    const output = testInput(input)
    expect(output).toMatchSnapshot()
  })

  it('should understand that strings written after <START>, up until the first name+colon, should be a system message.', () => {
    const input = neat`Sam: hey
      Vader: hi!
      <STaRT>
      test
      Vader: bye Sam
      Sam: byebye Vader
      Vader: I love you Sam
      Sam: me too Vader
      Vader is very excited.`
    const output = testInput(input)
    expect(output).toMatchSnapshot()
  })

  it('will trim result into budget', () => {
    const input = neat`
      Sam: Hey
      Vader: Hi!
      <START>
      test
      Vader: This is how I talk
      Sam: Oh, interesting!
      Vader: I also talk like this! *smiles* But this is far too long to include in our budget!
      Sam: More interesting!`
    expect(testInput(input, 25)).to.matchSnapshot()
  })

  it('should correctly separate out system messages it is explicitly handed, such as a post-sample marker string', () => {
    const input = neat`Sam: hey
      Vader: hi!
      <START>
      Vader: bye Sam
      Sam: byebye Vader
      System: New conversation started. Previous conversations are examples only.`
    const output = testInput(input)
    expect(output).toMatchSnapshot()
  })
})

function testInput(input: string, budget?: number) {
  const result = splitSampleChat({
    sampleChat: input,
    char: TEST_CHARACTER_NAME,
    sender: TEST_USER_NAME,
    budget,
  })

  return JSON.stringify(result.additions, null, 2)
}

const TEST_CHARACTER_NAME = 'Vader'
const TEST_USER_NAME = 'Sam'
