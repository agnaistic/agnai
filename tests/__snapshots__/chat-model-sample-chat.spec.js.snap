// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Completion Example Dialogue:: should correctly separate out system messages it is explicitly handed, such as a post-sample marker string 1`] = `
"[
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi!\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: bye Sam\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: byebye Vader\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: should properly convert <START> (case insensitive) and split basic messages 1`] = `
"[
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey there Vader!\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi Sam!\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: should properly understand the first conversation even if it doesnt begin with <START> 1`] = `
"[
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey there\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi!\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: should understand that strings written after <START>, up until the first name+colon, should be a system message. 1`] = `
"[
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi!\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"test\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: bye Sam\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: byebye Vader\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: I love you Sam\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: me too Vader\\\\nVader is very excited.\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: should understand when multiple <Start> are used in the sample chat and turn them all into the defined System message saying a new conversation has started 1`] = `
"[
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"Vader is nice.\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi!\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: bye Sam\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: byebye Vader\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: will interpret everything up to the first name+colon (or <start>) as a system message if the sampleChat string doesnt begin with a name+colon 1`] = `
"[
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"Vader is nice.\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi!\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: will interpret everything up to the first name+colon (or <start>) as a system message if the sampleChat string doesnt begin with a name+colon 2`] = `
"[
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"Vader is nice.\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: hey\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: hi!\\"
  }
]"
`;

exports[`Chat Completion Example Dialogue:: will trim result into budget 1`] = `
"[
  {
    \\"role\\": \\"user\\",
    \\"content\\": \\"Sam: Hey\\"
  },
  {
    \\"role\\": \\"assistant\\",
    \\"content\\": \\"Vader: Hi!\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"New conversation started. Previous conversations are examples only.\\"
  },
  {
    \\"role\\": \\"system\\",
    \\"content\\": \\"test\\"
  }
]"
`;
